realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -349,7 +351,34 @@ public void downloadAllServerChanges() throws InterruptedException {
         // In Java we cannot lock on the Session object either since it will prevent any attempt at modifying the
         // lifecycle while it is in a waiting state. Thus we use a specialised mutex.
         synchronized (waitForChangesMutex) {
-            waitForChanges(DIRECTION_DOWNLOAD);
+            waitForChanges(DIRECTION_DOWNLOAD, -1, null);
+        }
+    }
+
+    /**
+     * Calling this method will block until all known remote changes have been downloaded and applied to the Realm
+     * or the specified timeout was hit. This will involve network access, so calling this method should only be done
+     * from a non-UI thread.
+     * <p>
+     * This method cannot be called before the session has been started.
+     *
+     * @throws IllegalStateException if called on the Android main thread.
+     * @throws InterruptedException if the timeout was hit or the thread was interrupted while downloading was in progress.
+     * @throws IllegalArgumentException if {@code timeout} is less than {@code 0} or {@code unit} is {@code null}.
+     * @return {@code true} if the data was downloaded before the timeout. {@code false} if the operation timed out or otherwise failed.
+     */
+    public boolean downloadAllServerChanges(long timeout, TimeUnit unit) throws InterruptedException {
+        checkIfNotOnMainThread("downloadAllServerChanges() cannot be called from the main thread.");
+        checkTimeout(timeout, unit);
+
+        // Blocking only happens at the Java layer. To prevent deadlocking the underlying SyncSession we register
+        // an async listener there and let it callback to the Java Session when done. This feels icky at best, but
+        // since all operations on the SyncSession operate under a shared mutex, we would prevent all other actions on the
+        // session, including trying to stop it.
+        // In Java we cannot lock on the Session object either since it will prevent any attempt at modifying the
+        // lifecycle while it is in a waiting state. Thus we use a specialised mutex.
+        synchronized (waitForChangesMutex) {
```

Comment:
Since the lock is at the Java level only, then I question the usefulness of these overloads, since the `m_state_mutex` will still be acquired from the call to the OS's `session->wait_for_download_completion`, by returning you give the user the false impression that the underlying operation has been cancelled/timed out which  is clearly not the case. Imagine this use case 

```Java
SyncSession sync;
sync.downloadAllServerChanges(1, TimeUnit.MINUTES)
// do some commits 
sync.uploadAllServerChanges(1, TimeUnit.MINUTES);
// by returning here you have no guarantee that the actcual upload was triggered (at OS level), since we might still be holding the lock at the C++ level, from the downloadAll call
```

Reply:
We are using the asynchronous methods in the C++ layer, so no locks are actually being held. We are just using the Java lock to wait for the callback to happen. The reason for this that is that it is otherwise impossible to interrupt the Java thread at all, which would be highly undesirable.

IMO the semantics we have is that if `downloadAllServerChanges` completes, then changes have been downloaded. If it fails or times out, the behavior is undefined, i.e. it doesn't mean that downloading is canceled, it just means it didn't complete in time. I think that is acceptable. Behaving otherwise indicate a level of control over the syncing process that we don't offer (as everything happens more or less automatically).

Reply:
I have updated the JavaDoc to indicate that downloading/uploading might still happen.